### PR TITLE
Fine-grained kernel profiling

### PR DESCRIPTION
**Pull request by naoyam**
This instruments generated kernels to produce fine-grained performance profiles. Currently, only `GridReduction` and `GroupedGridReduction` are supported.

```
(pytorch-debug1) naoya-srv:debug1$ PYTORCH_NVFUSER_PROFILE_KERNEL=1 ./build/bin/test_jit  --gtest_filter="NVFuserTest.*GridReduction1_*"
Only one CUDA device detected. Disabling MultiCUDA tests
Note: Google Test filter = NVFuserTest.*GridReduction1_*-*_MultiCUDA
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from NVFuserTest
[ RUN      ] NVFuserTest.FusionGridReduction1_CUDA
Kernel performance profile:
GridReduction, T1, 3.267 us, 1
[       OK ] NVFuserTest.FusionGridReduction1_CUDA (2027 ms)
[----------] 1 test from NVFuserTest (2027 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2028 ms total)
[  PASSED  ] 1 test.
```

As you can see, after each kernel call, the time spent for each grid reduction is printed with its tensor name as well as the number of calls. The time is the average of all calls from the same call site. 

Internally, when enabled with the env var, an integer tensor is added to the kernel as a buffer to store profiling results. Time measurement is meant to be done using `clock64()`, and it currently needs to be manually inserted (so, only the runtime functions are supported). See the profiled version of `gridReduce` in `grid_reduction.cu`. 

Warnings
- Profiling DOES perturb the kernel performance itself, so the results should be used as an approximation. 
- As the cycle counter is used, timing results can be fluctuated if the GPU clock is not locked. On my system, I do `nvidia-smi -i 0 -ac 7001,1350`.

To Do for follow-up PRs
- Support welford
- Support measurement of ForLoops

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu [hunk 1]
```diff
@@ -294,6 +295,72 @@ __device__ void gridReduce(
...
+    bool X_BLOCK,
+    bool Y_BLOCK,
+    bool Z_BLOCK,
+    bool X_THREAD,
+    bool Y_THREAD,
+    bool Z_THREAD,
+    bool PERSISTENT_REDUCTION,
+    typename T,
+    typename Func>
+__device__ void gridReduce(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu:313]: Is there a cleaner mechanism in C++ to forward all arguments except the last two while also using template forwarding? I'm guessing maybe no, but might be cleaner.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu:313]: I did try `std::forward`, but it's not available with nvrtc (haven't tried with nvcc). We could probably develop something similar to `std::forward` by ourselves, but I'm not really confident especially in the device side with nvrtc.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/grid_reduction.cu:313]: Okay, doesn't seem worth it.

File: torch/csrc/jit/codegen/cuda/utils.cpp [hunk 2]
```diff
@@ -249,6 +249,11 @@ bool useFallback() {
   return !fallback_disabled;
 }
 
+bool isKernelProfileEnabled() {
+  const static auto env = getenv("PYTORCH_NVFUSER_PROFILE_KERNEL");
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/utils.cpp]: Can you put this under `PYTORCH_NVFUSER_ENABLE` interface Xiang has in https://github.com/csarofeen/pytorch/pull/1667 ?

File: torch/csrc/jit/codegen/cuda/lower_instrument.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  void allocateBuffer() {
+    const auto num_profile_entries = profile_.getNumberOfProfileEntries();
+
+    // If nothing to profile, do not allocate anything
+    if (num_profile_entries == 0) {
+      return;
+    }
+
+    // Allocate two integers for each etnry. One is used for accumulating
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_instrument.cpp]: Nit: `entry`

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 4]
```diff
@@ -358,6 +361,73 @@ std::vector<Expr*>& KernelInternalProxy::topLevelExprs() {
...
+  }
+
+  auto slot = getNewIndex();
+  expr_entry_map_.emplace(expr, slot);
+}
+
+int KernelPerformanceProfile::getNewIndex() {
+  auto slot = num_profile_entries_;
+  ++num_profile_entries_;
+  return slot;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: Nit: isn't this just `return num_profile_entries_++;`

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 5]
```diff
@@ -358,6 +361,73 @@ std::vector<Expr*>& KernelInternalProxy::topLevelExprs() {
...
+    return c10::optional<int>();
+  } else {
+    return it->second;
+  }
+}
+
+std::array<int, 2> KernelPerformanceProfile::getIndicesInProfileBuffer(
+    const Expr* expr) const {
+  if (!isProfiled(expr)) {
+    return {};
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: Should you assert if it's not profiled? I assume std::array will initialize to `0, 0` here?

File: torch/csrc/jit/codegen/cuda/lower_instrument.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//! Set up kernel profiling when enabled
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_instrument.h]: Add some more to the comment here describing the pass a bit please. Some good notes to add:
New Exprs are returned as normal lowering passes tend to do. The buffer allocation to record profiling information is added to the beginning of the Exprs. However the big purpose of this lowering pass is to fill the information about the Expr's to be profiled in KernelPerformanceProfile which escapes this scope through the current GpuLower object.

**Review by csarofeen — APPROVED** [review]
Before grabbing the initial clock count, should we thread fence? Since it is asynchronous computing it seems strange to me that we don't do some sort of synchronization before grabbing the start time.

Otherwise looks good to me, just some nit's here and there.

File: torch/csrc/jit/codegen/cuda/utils.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/utils.cpp]: Done

File: torch/csrc/jit/codegen/cuda/kernel.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel.cpp]: Yes. I think I was doing something more but ended up just with this. Cleaned it up.

File: torch/csrc/jit/codegen/cuda/kernel.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel.cpp]: Assertion seems more reasonable. Updated.

File: torch/csrc/jit/codegen/cuda/lower_instrument.h (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_instrument.h]: Thanks. Added more comments.

**Comment by naoyam** [comment]
> Before grabbing the initial clock count, should we thread fence? Since it is asynchronous computing it seems strange to me that we don't do some sort of synchronization before grabbing the start time.
> 
> Otherwise looks good to me, just some nit's here and there.

`__threadfence()` seems to make sense. I added a function to do `__threadfence()` and then `clock64()`, which is used instead of directly calling `clock64()`. This would make it easier to change the behavior if necessary.

**Comment by naoyam** [comment]
Maybe we should also consider using inline PTX.

https://forums.developer.nvidia.com/t/cost-of-clock64/38497/15

**Pull request merged**